### ISSUE TITLE
Error Installing FlashAttention on Windows 11 with CUDA 11.8 - "CUDA_HOME environment variable is not set"

### ISSUE THREAD

**Issue by Mr-Natural** [Dao-AILab/flash-attention#982]
I am encountering an error while attempting to install the flash-attn library on my Windows 11 machine with CUDA 11.8. Despite having the nvcc compiler and CUDA properly installed and accessible, the installation fails due to a supposed missing CUDA_HOME environment variable.

Setup Information:
Operating System: Windows 11 (version 10.0.22631.3593)
CUDA Version: 11.8
PyTorch Version: 2.3.0+cpu
NVIDIA Driver Version: 522.06

Steps to Reproduce:

1. Set CUDA_HOME environment variable:
setx CUDA_HOME "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.8"
setx PATH "%CUDA_HOME%\bin;%PATH%"
setx PATH "%CUDA_HOME%\lib64;%PATH%"

2. Verify CUDA_HOME and nvcc
echo %CUDA_HOME%
nvcc --version

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Wed_Sep_21_10:41:10_Pacific_Daylight_Time_2022
Cuda compilation tools, release 11.8, V11.8.89
Build cuda_11.8.r11.8/compiler.31833905_0

3. Create and activate a new Conda environment:
conda create -n phivision_clean python=3.9
conda activate phivision_clean

4. Install prerequisites:
pip install packaging ninja
pip cache purge
set MAX_JOBS=4

5. Attempt to install FlashAttention
pip install flash-attn --no-build-isolation

Error Log:
Collecting flash-attn
  Downloading flash_attn-2.5.9.post1.tar.gz (2.6 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.6/2.6 MB 5.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [22 lines of output]
      fatal: not a git repository (or any of the parent directories): .git
      C:\Users\allan\AppData\Local\Temp\pip-install-9pn3ykxp\flash-attn_1cfbe659b3424b648b092c7a32f1614f\setup.py:78: UserWarning: flash_attn was requested, but nvcc was not found.  Are you sure your environment has nvcc available?  If you're installing within a container from https://hub.docker.com/r/pytorch/pytorch, only images whose names contain 'devel' will provide nvcc.
        warnings.warn(
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "C:\Users\allan\AppData\Local\Temp\pip-install-9pn3ykxp\flash-attn_1cfbe659b3424b648b092c7a32f1614f\setup.py", line 134, in <module>
          CUDAExtension(
        File "C:\Users\allan\miniconda3\envs\phivision\Lib\site-packages\torch\utils\cpp_extension.py", line 1077, in CUDAExtension
          library_dirs += library_paths(cuda=True)
                          ^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\allan\miniconda3\envs\phivision\Lib\site-packages\torch\utils\cpp_extension.py", line 1211, in library_paths
          paths.append(_join_cuda_home(lib_dir))
                       ^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\allan\miniconda3\envs\phivision\Lib\site-packages\torch\utils\cpp_extension.py", line 2419, in _join_cuda_home
          raise OSError('CUDA_HOME environment variable is not set. '
      OSError: CUDA_HOME environment variable is not set. Please set it to your CUDA install root.

      torch.__version__  = 2.3.0+cpu

      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

Additional Information:

nvcc output:
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Wed_Sep_21_10:41:10_Pacific_Daylight_Time_2022
Cuda compilation tools, release 11.8, V11.8.89
Build cuda_11.8.r11.8/compiler.31833905_0

nvidia-smi output:
Sat Jun  8 21:58:06 2024
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 522.06       Driver Version: 522.06       CUDA Version: 11.8     |
|-------------------------------+----------------------+----------------------+
| GPU  Name            TCC/WDDM | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ... WDDM  | 00000000:01:00.0  On |                  N/A |
| 31%   35C    P8    13W / 175W |   1125MiB /  8192MiB |      2%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

Notes:
Despite setting CUDA_HOME and confirming nvcc is available, the installation script does not seem to recognize the CUDA_HOME variable.
The error might also be linked to residual files or incorrect environment setup from previous installation attempts.
Thank you for your assistance.

**Comment by Tellterubouzu** [comment]
This may be unrelated, but is your torch the cpu version instead of the cuda?

**Comment by Mr-Natural** [comment]
Thanks for the question. The version I am using is GPU enabled one. Details: torch.__version__  = 2.3.1+cu118

**Comment by bdashore3** [comment]
FA2 does not support cuda 11.8 on windows. You'd need to update to cuda 12 and install a prebuilt wheel from my fork https://github.com/bdashore3/flash-attention/releases

**Comment by Mr-Natural** [comment]
Thank you for taking the time to respond. I'll follow your guidance and all going to plan things will be operational soon!

**Comment by zymox** [comment]
> FA2 does not support cuda 11.8 on windows. You'd need to update to cuda 12 and install a prebuilt wheel from my fork https://github.com/bdashore3/flash-attention/releases

I installed CUDA 12.2 and installed the fitting wheel from your repo.
However, just running 
```python
import flash_attn
print(flash_attn.__version__)
``` 
I get the error that **flash_attn_2_cuda** module cannot be found
```
Traceback (most recent call last):
  File "C:\Users\tuete\PycharmProjects\Florence-2-large-ft\testing.py", line 1, in <module>
    import flash_attn
  File "C:\Dev\Python3.10\lib\site-packages\flash_attn\__init__.py", line 3, in <module>
    from flash_attn.flash_attn_interface import (
  File "C:\Dev\Python3.10\lib\site-packages\flash_attn\flash_attn_interface.py", line 10, in <module>
    import flash_attn_2_cuda as flash_attn_cuda
ImportError: DLL load failed while importing flash_attn_2_cuda: Das angegebene Modul wurde nicht gefunden.
``` 

Any idea how to fix that?

**Comment by zymox** [comment]
Nevermind, it was because I had the wrong torch version (without CUDA available), it worked after uninstalling torch and reinstalling with the command from https://pytorch.org/get-started/locally/
```
pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121
```

**Comment by Mr-Natural** [comment]
Thanks for the reply @zymox

**Comment by KOG-Nisse** [comment]
I have cuda 12.5 installed. Is this version compatible?

**Comment by Swinster6** [comment]
> I have cuda 12.5 installed. Is this version compatible? ... Update: it worked

Hey! How did you get it to work? I have cuda 12.5 too.

**Comment by KOG-Nisse** [comment]
> > I have cuda 12.5 installed. Is this version compatible? ... Update: it worked
> 
> Hey! How did you get it to work? I have cuda 12.5 too.

I downloaded and ran the wheel.
https://github.com/bdashore3/flash-attention/releases

Also, make sure you have the right version of cuda installed:
pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu124
https://pytorch.org/get-started/locally/

**Comment by Rahman2001** [comment]
@KOG-Nisse Thank you! That solved my problem. I think we can close this issue.

**Comment by Rahman2001** [comment]
First, download a correct wheel from the link you mentioned. If you don't know how to install it, follow instructions in this link: [in stackoverflow](https://stackoverflow.com/a/27909082/15605796)

**Comment by ucaokylong** [comment]
> First, download a correct wheel from the link you mentioned. If you don't know how to install it, follow instructions in this link: [in stackoverflow](https://stackoverflow.com/a/27909082/15605796)

I check into your wheel link and see the package for cuda 12.3, but my cuda is 12.1 , is it compatible?

**Comment by tridao** [comment]
Yes

**Comment by anshumankmr** [comment]
I don't have a GPU, how do I run this?

**Comment by tridao** [comment]
> I don't have a GPU, how do I run this?

This code is for GPU.

**Comment by monarchwadia** [comment]
TL;DR run `python -c "import torch; print(torch.cuda._is_compiled())"`. If it returns `False`, then you have to create a new conda environment, install Torch from scratch from the official website, and then try installing flash-attn.

---

Turns out that `CUDA_HOME` does not refer to an environment variable at all! Instead, it refers to a constant inside Torch code!

My stack trace had the following:

```bash
        File "C:\Users\myname\miniconda3\envs\librarian\Lib\site-packages\torch\utils\cpp_extension.py", line 2416, in _join_cuda_home
          raise OSError('CUDA_HOME environment variable is not set. '
      OSError: CUDA_HOME environment variable is not set. Please set it to your CUDA install root.
```

I stepped into the `cpp_extension.py` file and found the following line.

```python
CUDA_HOME = _find_cuda_home() if torch.cuda._is_compiled() else None
```

Wow. That's interesting. So the first thing it's checking for is `torch.cuda._is_compiled()`. I figured I should check for it. So I then ran the following command

```
python -c "import torch; print(torch.cuda._is_compiled())"
```

Sure enough, the output was `False`

I wiped my `conda` environment completely and created a new environment.

```
conda deactivate
conda env create -n some_environment_name
```

Then I went to Pytorch's website, https://pytorch.org/get-started/locally/, and ran the necessary commands to install Pytorch.


Finally, after running the command (I chose `conda install` instead of `pip`), I ran

```
pip install flash-attn
```

And then again ran the command to check

```
python -c "import torch; print(torch.cuda._is_compiled())"
```

This time it returned `True`

Wild ride!

**Comment by GarbageHaus** [comment]
After about 6+ hours of filling around trying to get this installed, it appears to be only installable on a Linux machine. It does say in the repo that only Linux is supported, so if you were looking to install this on Windows 10 you will not be able to.

I was only able to get as far as I did by modifying the attention.py file. See: https://github.com/Dao-AILab/flash-attention/issues/509

As above, running: `python -c "import torch; print(torch.cuda._is_compiled())"` did return "True" however, this did not prevent the issue of it not recognizing flash_attn_2_cuda.

Doing a deep dive I found I needed to manually install the full CUDA kit from Nvidia's website (No other method would work) just to get it to run through the setup.py.

Then it required MS Build Tools. (I didn't see this documented on the repo.)

Ultimately it still failed due to "Error" when calling _run_ninja_build from torch's ccp_extension.py. The error, so far, as been too vague for me to understand.

So as far as I can determine, this repo cannot be installed via any method on a Windows 10 machine.

**Comment by AlanRogersScripts** [comment]
> After about 6+ hours of filling around trying to get this installed, it appears to be only installable on a Linux machine. It does say in the repo that only Linux is supported, so if you were looking to install this on Windows 10 you will not be able to.
> 
> I was only able to get as far as I did by modifying the attention.py file. See: [#509](https://github.com/Dao-AILab/flash-attention/issues/509)
> 
> As above, running: `python -c "import torch; print(torch.cuda._is_compiled())"` did return "True" however, this did not prevent the issue of it not recognizing flash_attn_2_cuda.
> 
> Doing a deep dive I found I needed to manually install the full CUDA kit from Nvidia's website (No other method would work) just to get it to run through the setup.py.
> 
> Then it required MS Build Tools. (I didn't see this documented on the repo.)
> 
> Ultimately it still failed due to "Error" when calling _run_ninja_build from torch's ccp_extension.py. The error, so far, as been too vague for me to understand.
> 
> So as far as I can determine, this repo cannot be installed via any method on a Windows 10 machine.

I am struggling with the same issue and not having much luck. If you go to their releases it does seem to suggest that there are versions for "Win"

https://github.com/kingbri1/flash-attention/releases

**Comment by supreme-gg-gg** [comment]
I'm trying to build a docker image with flash attention 2 and faced the same issue:

```
solved 26 packages in 1.09s
Step #1:    Building flash-attn==2.8.0.post2
Step #1:   × Failed to build `flash-attn==2.8.0.post2`
... 
line 2525, in _join_cuda_home
Step #1:           raise OSError('CUDA_HOME environment variable is not set. '
Step #1:       OSError: CUDA_HOME environment variable is not set. Please set it to
```

This is my line from dockerfile that installs dependencies (using Python 3.12 slim base image):

```
RUN uv venv && \
    uv pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu124 && \
    uv sync --frozen --no-cache --verbose && \
    uv pip install flash-attn --no-build-isolation
```

Seems like this thread is mostly for Windows but if anyone has any idea would appreciate!

**Comment by velanteg** [comment]
Why CUDA 13 not supported everywhere?
